### PR TITLE
Pidfd based forking with supervisor

### PR DESCRIPTION
**Pull request by mensfeld**
This PR provides ability to:
1. Fork during start
2. Recover fork after child crashes
3. Parent death + shutdown of orphans
4. In sync cluster operations (shutdown, quiet)
5. Forceful children shutdown after timeout
6. Health checking API via pipe
7. CLI command `bundle exec karafka swarm` with same options as `bundle exec karafka server`
8. Additionally dependency injection was added to simplify configuration injection

Things still missing (for v1):
1. [x] Ability to configure number of forks
2. [x] New config options validations
3. [x] Unit specs
4. [x] Integration tests
5. [x] Logger Listener updates
6. [x] K8s listener update (swarm listener introduced)
7. [x] Eager load hooks
8. [x] Failure on systems not supporting pidfd and fork  

1.5 year :sweat_smile: 

close https://github.com/karafka/karafka/issues/1058

### CODE REVIEW

**Review by KJTsanaktsidis — COMMENTED** [review]
(oh sorry i wrote these comments yesterday but I didn't actually submit the review!)

File: lib/karafka/swarm/node.rb [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+          monitor.subscribe(liveness_listener)
+          monitor.instrument('swarm.node.after_fork', caller: self)
+
+          Server.run
+
+          @writer.close
+        end
+
+        @writer.close
+        @pidfd = Pidfd.new(@pid)
```
[the remark this answers is not in the document]
Comment (mensfeld) [reply lib/karafka/swarm/node.rb:78]: No but I actually missed the sigchld handler trigger for post-state change management. Thanks!

File: lib/karafka/swarm/node.rb [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+      end
+
+      # Indicates that this node is doing well
+      # @note Child API
+      def healthy
+        write(1)
```
[the remark this answers is not in the document]
Comment (mensfeld) [reply lib/karafka/swarm/node.rb]: isn't `read_nonblock` anyhow returning it as a string?

```ruby
irb(main):016> a,b = IO.pipe
=> [#<IO:fd 7>, #<IO:fd 8>]
b.write_nonblock(0x01)
b.write_nonblock(0x01)
b.write_nonblock(0x01)
a.read_nonblock(1024)
"111"
```

I understand the confusion though, will adjust

File: lib/karafka/swarm/pidfd.rb [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+      def alive?
+        @pidfd_select ||= [@pidfd_io]
+
+        IO.select(@pidfd_select, nil, nil, 0).nil?
+      end
+
+      # Cleans the zombie process
+      # @return [Boolean] true if collected, false if process is still alive
+      def cleanup
+        !::Process.waitpid(@pid, ::Process::WNOHANG).nil?
```
[the remark this answers is not in the document]
Comment (mensfeld) [reply lib/karafka/swarm/pidfd.rb:71]: Great point!

File: lib/karafka/swarm/pidfd.rb [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        # @return [Boolean] true if syscall is supported via FFI
+        def supported?
+          # If we were not even able to load the FFI C lib, it won't be supported
+          return false unless API_SUPPORTED
+          # Won't work on macOS because it does not support pidfd
+          return false if RUBY_DESCRIPTION.include?('darwin')
+          # Won't work on Windows for the same reason as on macOS
+          return false if RUBY_DESCRIPTION.match?(/mswin|ming|cygwin/)
+
+          # There are some OSes like BSD that will have C lib for FFI bindings but will not support
```
Comment (KJTsanaktsidis) [inline lib/karafka/swarm/pidfd.rb:51]: Or if you're using a new glibc on an old kernel too.

File: lib/karafka/swarm/pidfd.rb [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+      # @return [Boolean] true if signal was sent, otherwise false or error raised. `false`
+      #   returned when we attempt to send a signal to a dead process
+      # @note It will not send signals to dead processes
+      def signal(sig_name)
+        @mutex.synchronize do
+          return false if @cleaned
+          # Never signal processes that are dead
+          return false unless alive?
+
+          result = fdpid_signal(
```
Comment (KJTsanaktsidis) [inline lib/karafka/swarm/pidfd.rb:99]: yeah using `pidfd_send_signal` is good, no more races!

**Review by nijikon — APPROVED** [review]
An epic piece of engineering. Another great feature of the Karafka ecosystem :purple_heart:

**Pull request merged**